Optimize Workflow Caching

ObliviousHarmony opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

This pull request centralizing the dependency installation, building, and caching within a single composite action. Since virtually every single workflow runs both `cache-deps` and `install-build`, it seems reasonable that these should instead be encapsulated in a single action. This pull request also adds support for caching build outputs with the intention of dramatically speeding up related CI checks.

Closes #33908.

### How to test the changes in this Pull Request:

1. Confirm that all CI checks are passing in this PR.
2. Confirm that the "Setup WooCommerce Monorepo" steps in all workflows pulls from the cache successfully.

Once you've confirmed that things seem to be working here, there's some steps you should take in a fork of `woocommerce/woocommerce`.

1. Confirm that the first run of CI does not populate from cache. The "Setup WooCommerce Monorepo" step should take a while to run.
2. Confirm that the second run of CI populates from the cache. The "Setup WooCommerce Monorepo" step should run quickly.
3. Merge the contents of this PR into your fork.
4. Verify that the "Prime Cache" workflow runs correctly
5. Create a new PR making a change to WooCommerce's PHP files in a commit in your fork. Confirm that a minimal install and rebuild takes place based on the contents of the PR. This should work because the cache priming in `trunk` saves the day.

Once you've confirmed that the workflows function correctly, you should play around with the inputs on the action to ensure that they work in different cases.

1. `install-filters` supports both single and multi-line filters. One filter per line. This is passed to `pnpm install` and helps indicate what should be installed in the monorepo. This is nice for smaller installations.
2. `build` supports true or false and should skip the Turbo build step if it is false.
3. `build-filters` supports both single and multi-line filters. One filter per line. This is passed to Turborepo's build and is nice to limit the building to a single package or set of packages.
4. `php-version` supports changing the version of PHP that the action sets up.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

ObliviousHarmony committed:
Added Monorepo Setup Action

This action will handle the installation, building, and caching for all
projects within the monorepo. It has inputs for skipping builds and
filtering so that only specific packages are installed and built.

ObliviousHarmony committed:
Testing Workflow

ObliviousHarmony committed:
Workflow Test

ObliviousHarmony committed:
Test Workflow

ObliviousHarmony committed:
Fixed Default Argument Quoting

ObliviousHarmony committed:
Fixed Missing Shell Input

ObliviousHarmony committed:
Fixed Missing PNPM Command

ObliviousHarmony committed:
Fixed Composer Cache Directory

psealock commented on .github/actions/setup-woocommerce-monorepo/action.yml:
```diff
@@ -0,0 +1,56 @@
+  using: composite
+  steps:
+    - name: Setup PNPM
+      uses: pnpm/action-setup@10693b3829bf86eb2572aef5f3571dcf5ca9287d
```
Any particular reason to pin to a commit? If so, a comment here and below would be helpful

ObliviousHarmony replied:
[GitHub's documentation on security hardening](https://github.com/woocommerce/woocommerce/pull/34607) encourages maintainers to pin commit hashes for actions in order to avoid the possibility of malicious code being shipped in an action. While it does make maintenance a bit more difficult, this is likely something we should be doing in _all_ of our actions.

samueljseay replied:
I guess doing this will also encourage us to make actions more modular too so its win/win :D

psealock reviewed (commented):
>Confirm that the cache is populated in subsequent executions of the workflow

Do you mean subsequent executions of the the matrix in a single PR, or if I rerun the entire workflow?

ObliviousHarmony committed:
Removed Test Caching

Caching the results of tests, while an interesting way to avoid
unnecessary execution, seems error-prone. We can't adequately
capture the environment such that we can rely on this working.
For instance, changes in PHP version might break PHPUnit tests,
but, cached runs will be shared across different workflows.

It seems better to just not cache it and rely on people only testing
code that has actually changed in the pull request.

ObliviousHarmony committed:
Added Input Parsing

This SHOULD allow for passing multiple filters as an array argument.

ObliviousHarmony committed:
Fixed Outputs

ObliviousHarmony committed:
Fixed Input Parsing

ObliviousHarmony committed:
Fixed Input

ObliviousHarmony committed:
Fixed Inputs

ObliviousHarmony committed:
Fix Input Parsing

ObliviousHarmony committed:
Escaped Filter Input

ObliviousHarmony committed:
Added Default

ObliviousHarmony committed:
Fixed Defaults

ObliviousHarmony committed:
Fixed Input

ObliviousHarmony committed:
Removed Default

ObliviousHarmony committed:
Added Filter Default

ObliviousHarmony committed:
Fixed Install Actions

ObliviousHarmony committed:
Testing Filters Input

ObliviousHarmony committed:
Separated Install and Build filters

ObliviousHarmony committed:
Testing Multiline FIlter

ObliviousHarmony committed:
Added build disabling

ObliviousHarmony committed:
Changed Bash To Node

This is simpler and will stop
all of my headaches.

ObliviousHarmony committed:
Test multiple filters

ObliviousHarmony committed:
Too excited to be done and missed quote

ObliviousHarmony committed:
Cleaned up filter generation

ObliviousHarmony committed:
Test build disabling

ObliviousHarmony committed:
Fixed Unit Test Workflow

ObliviousHarmony committed:
Updated Workflows

This fully utilizes the new action across all of the workflows.

ObliviousHarmony committed:
Always Cache Build Output

Since there may be other commands run that populate the cache, we
should make sure that we are always caching just in case.

ObliviousHarmony committed:
Fixed Code Coverage

samueljseay reviewed (approved):
This tests well and is a huuuge improvement to build speed. 🚀
